src/daft-dsl/src/expr.rs

```diff
@@ -1174,6 +1181,106 @@ fn expr_has_agg(expr: &ExprRef) -> bool {
     }
 }
 
+// Finds the names of all the wildcard expressions in an expression tree.
+// Needs the schema because column names with stars must not count as wildcards
+fn find_wildcards(expr: ExprRef, struct_expr_map: &HashMap<String, ExprRef>) -> Vec<Arc<str>> {
+    match expr.as_ref() {
+        Expr::Column(name) => {
+            if name.contains('*') {
+                if struct_expr_map.contains_key(name.as_ref()) {
+                    log::warn!(
+                        "Warning: Column '{name}' contains *, preventing potential wildcard match"
+                    );
+                    Vec::new()
+                } else {
+                    vec![name.clone()]
+                }
+            } else {
+                Vec::new()
+            }
+        }
+        _ => expr
+            .children()
+            .into_iter()
+            .flat_map(|e| find_wildcards(e, struct_expr_map))
+            .collect(),
+    }
+}
+
+// Calculates a list of all wildcard matches against a schema.
+fn get_wildcard_matches(
+    pattern: &str,
+    schema: &Schema,
+    struct_expr_map: &HashMap<String, ExprRef>,
+) -> DaftResult<Vec<String>> {
+    if pattern == "*" {
+        // return all top-level columns
+        return Ok(schema.fields.keys().cloned().collect());
+    }
+
+    if !pattern.ends_with(".*") {
+        return Err(DaftError::ValueError(format!(
+            "Unsupported wildcard format: {pattern}"
+        )));
+    }
+
+    // remove last two characters
+    let mut struct_name = pattern.to_string();
+    struct_name.pop();
+    struct_name.pop();
+
+    let Some(struct_expr) = struct_expr_map.get(&struct_name) else {
+        return Err(DaftError::ValueError(format!(
+            "Error matching wildcard {pattern}: struct {struct_name} not found"
+        )));
+    };
+
+    // get innermost struct from expr
+    let field = struct_expr.to_field(schema)?;
+    let DataType::Struct(subfields) = field.dtype else {
+        return Err(DaftError::ValueError(format!(
+            "Error matching wildcard {pattern}: {struct_name} has dtype {}, not struct",
+            field.dtype
+        )));
```

Comment:
I think have a possible edge case to this. What if your schema was: 
```
{
    "a.b": int
    "a": {
        "b": {
            "c": int
        }
    }
}
```
In that case, "a.b.*" would match to the field "a.b", which would error since it's not a struct, but there is indeed a valid matching of this wildcard in the schema. 

I think the solution might be for `calculate_struct_expr_map` to compute an ordered vec of possible expressions for each string instead of just the first one, and then here you would go down that vec and find the first expression that is a struct.

Reply:
Fixed by having `calculate_struct_expr_map` return all possible matched expressions in order, instead of just one.

This also means it is now possible to tell if a struct get has multiple interpretations. Maybe we should log a warning if this happens?

Reply:
We could, feel free to add it if you'd like